### PR TITLE
move expected_loglik here from ApproximateGPs.jl

### LINKED ISSUE #71: Precompute the nodes and weights for `GaussHermite`
Right now the nodes and weight are recomputed every time `expected_loglik` is called. The constructor of `GaussHermite` could simply store the two as vectors.

### PR DESCRIPTION
**Pull request by st--**
See #42 - ~should discuss the design while we're at it!~ any further design discussions moved to #73

Main changes to what's currently in ApproximateGPs:

- rename to `expected_loglikelihood` to be more explicit
- changed signature: argument order is now `quadrature`, `lik`, `q_f`, `y`
- rename expectation structs to `DefaultExpectationMethod`, `AnalyticExpectation`, `GaussHermiteExpectation`, `MonteCarloExpectation`
- drop abstract base type (weren't used for anything)
- drop default argument for GaussHermite / MonteCarlo quadrature points (moved into `DefaultExpectationMethod`)
- resolves #71 in c25b2ce
- move analytic definitions into files for each likelihood implementation

Note: bumps julia compat to 1.6.

### CODE REVIEW

File: src/GPLikelihoods.jl
Comment (theogf) [reply src/GPLikelihoods.jl:31]: Because `Analytic` and `MonteCarlo` are technically not quadrature methods? (correct me if I am wrong)
To go with your remark `DefaultQuadrature` should be called `DefaultIntegrationMethod` I guess.

I don't see the point of going with the SciML interface, cause we don't need one. The interface approaches two different problems. We have cases where we want to use sample / analytic and another where we want to use quadrature.
However I don't see the point of having different quadrature methods. Gauss-Hermite is already optimal given an expectation over a Gaussian distribution. (again correct me if it's incorrect)

Comment (theogf) [reply src/GPLikelihoods.jl:31]: But +1 on too generic names.
I guess adding an `Integration` postfix would solve the problem.

Comment (devmotion) [reply src/GPLikelihoods.jl:31]: > However I don't see the point of having different quadrature methods.

Well, in the sense of Quadrature.jl also eg Monte Carlo methods (and Quasi Monte Carlo methods: https://github.com/SciML/Quadrature.jl/issues/11) are "quadrature" algorithms. So I guess the main issue are not the algorithms but that Quadrature is written for problems of integrating a generic function `f`, without any decomposition into `g(x) * d(x)` where `d` is a density.

> Gauss-Hermite is already optimal given an expectation over a Gaussian distribution.

Optimal for a given tolerance? And optimal among Gauss quadrature methods, general quadrature rules, or general numerical integration algorithms?

Comment (st--) [reply src/GPLikelihoods.jl:31]: Yeah I agree the names are too generic, it's just what we had in ApproximateGPs before. Happy to change it before we merge it.

I don't remember having come across SciML/Quadrature, your suggestion might have simply slipped through some cracks, sorry! It would be great to re-use that, but if as theo says the interface doesn't quite work for us, then that's no good.

Iirc, nth-order Gauss-Hermite is exact for integration of nth-order polynomials under a Gaussian density. It's not necessarily optimal. It's been a while and I've never looked into it deep enough, but in GPflow context at some point I was discussing with someone about unscented transform(?) potentially being better than Gauss-Hermite.

Comment (theogf) [reply src/GPLikelihoods.jl:31]: Hmm unscented transform can definitely be better than Monte Carlo but I am not sure about Gauss-Hermite?

Comment (devmotion) [reply src/GPLikelihoods.jl:31]: I still don't know which kind of optimality you're referring to. I guess I've been involved in SciML for too long and hence expect from a numerical approximation method to be able to specify a desired accuracy a priori - like in DE solvers or the integration methods/packages supported by Quadrature. Optimality then just means fastest method for a given error, as illustrated in the work precision diagrams. And some methods are then eg better in the low tolerance regime and some in the high tolerance regime. I've used FastGaussQuadrature myself but this concept of having a fixed number of nodes without adapting it to your problem and desired precision goes against my numerical intuition and feels unsatisfying. And makes me wonder in what sense one can claim optimality.

(I'm fully aware though that it's nice and efficient to compute a fixed set of nodes upfront in eg an iterative optimization procedure. And it might be "good enough". But I'd like to have guarantees on "good enough" and use that to choose the smallest number of nodes for which the result is still "good enough".)

Comment (theogf) [reply src/GPLikelihoods.jl:31]: I think the discrepancy in our understanding is in the way we use quadrature. In GPs quadrature is just a small piece and we just want something as good as possible with the least possible computations and that's exactly what FGQ is doing.
Additionally, we need to solve the quadrature problem N times per step, so starting to care about tolerance might slow things down too much.
Maybe I have the (possibly wrong) view that the likelihoods used are always quite well-behaved and FGQ is always doing the trick for this problem.

Comment (devmotion) [reply src/GPLikelihoods.jl:31]: Yeah, as I wrote, it's nice that you only need one set of precomputed nodes.

> we just want something as good as possible with the least possible computations and that's exactly what FGQ is doing.

Yeah but that's exactly my point: how do you know it's good enough AND performant if you just use eg 200 nodes? Maybe 50 would have been sufficient and you could have been much more efficient. Or maybe you're result is not very accurate but you never know. The most natural approach would be to rerun computations with different number of nodes, possibly even optimizing the number of nodes - is this done in practice?

Comment (theogf) [reply src/GPLikelihoods.jl:31]: Yep! I agree it's bad but that's what most people do. It's not the only thing after all :) (# of inducing points etc).
That's also why I prefer augmentations :p : you don't need any quadrature :)

But again, it's generally not considered an issue because the integrand used in GPs are usually well-behaved and people just use something round like 100 nodes and are happy with it.

File: Project.toml
Comment (devmotion) [inline Project.toml:28]: BTW the Julia registry doesn't allow to bump the Julia compat in a patch release.

**Comment by st--** [comment]
Open question from #42:

> `expected_loglik` takes a Vector `y`, a Vector `q_f`, and a `lik` function that maps from scalar `f` to UnivariateDistribution; `expected_loglik` then does the broadcasting internally. Is that what we want?
>
>  - Would it be better (cleaner) to have `expected_loglik` handle scalars only (but this might result in some performance loss for e.g. recomputing the Gauss-Hermite points and weights for each data point)?
>  - Should we expect `lik` to take the full Vector `fs` and return e.g. a Product() of UnivariateDistributions (though this might make the expectation code more complicated)?
>  - How would we want to handle the heteroskedastic case, where we do want to include the correlations across the two outputs at each data point, but independence between different data points (not sure how we would handle that on the AbstractGPs side given that multi-output is all "output as input")?

Do we need to resolve that now, or should we merge it in this form and consider this later?

File: src/expectations.jl
Comment (st--) [inline src/expectations.jl:54]:
```julia
function expected_loglikelihood(quad, lik, q_f::AbstractMvNormal, y::AbstractVector)
    mean, std = mean_and_std(q_f)
    return expected_loglikelihood(quad, lik, Normal.(mean, std), y)
end
```

File: src/GPLikelihoods.jl
Comment (willtebbutt) [reply src/GPLikelihoods.jl]: @devmotion are you happy to resolve this now that @st-- has agreed that we don't need to export the names?

**Comment by willtebbutt** [comment]
> Do we need to resolve that now, or should we merge it in this form and consider this later?

Later. I'm not keen to hold this PR up. These points all require more design discussion and are part of larger problems. In my view they're best dealt with in the next breaking release.

File: src/expectations.jl
Comment (willtebbutt) [inline src/expectations.jl:59]:
```suggestion
The expected log likelihood, using the default quadrature method for the given likelihood.
(The default quadrature method is defined by `default_expectation_method(lik)`, and should
be the closed form solution if it exists, but otherwise defaults to Gauss-Hermite
quadrature.)
```
(line length)

File: src/expectations.jl
Comment (willtebbutt) [inline src/expectations.jl:37]:
```suggestion
where `p(y | f)` is the process likelihood. This is described by `lik`, which should be a
callable that takes `f` as input and returns a Distribution over `y` that supports
`loglikelihood(lik(f), y)`.
```
(line length)

File: test/expectations.jl
Comment (willtebbutt) [inline test/expectations.jl:50]:
```suggestion
    # see https://github.com/JuliaGaussianProcesses/ApproximateGPs.jl/issues/82
    @testset "testing Zygote compatibility with GaussHermiteExpectation" begin 
```
line length

File: test/expectations.jl
Comment (willtebbutt) [inline test/expectations.jl:76]:
```suggestion

        # Test GammaLikelihood parameter
```
readability

File: test/expectations.jl
Comment (willtebbutt) [inline test/expectations.jl:64]:
```suggestion

        # Test differentiation with likelihood parameters
```
readability

File: test/expectations.jl
Comment (willtebbutt) [inline test/expectations.jl:55]:
```suggestion

        # Test differentiation with variational parameters
```
readability

File: test/expectations.jl
Comment (willtebbutt) [inline test/expectations.jl:1]:
```suggestion
@testset "expectations" begin
```
match source file name

File: test/runtests.jl
Comment (willtebbutt) [inline test/runtests.jl:24]:
```suggestion
    include("expectations.jl")
```
not in separate folder, so fine to put by itself

**Review by willtebbutt — APPROVED** [review]
Nothing of substance to add, just style + formatting things that the formatter doesn't seem to have picked up on.

I'm happy for this to go in provided that @devmotion is satisfied that the currently-unresolved conversation has been resolved to his satisfaction.

**Pull request merged**
